### DISCUSSION TITLE
Import Best Practice

### DISCUSSION THREAD

**Discussion by XCame** [atrocore/atrocore#816] [Q&A]
Hi there, 

I am currently evaluating Atropim / Atrocore and so far I liked what I see. 

I now wanted to test it performance wise and wanted to import a csv file containing around 370k products. 
For the start I only tried to import 3 columns. GTIN, Name and a Color Attribute, that I imported beforehand from a short version of the large import file containing only like 20 colors (this worked flawlessly ) 

No I tried to import the 370k products. 

At first as a single import job without defining "Max per Job".
It worked, but extreme slow (maybe 100 per Minute). 

So I tried using MaxPerJob now with 2000 and I increased the worker count to 8.
Looked promising at first. with 2000 in 5-6 minutes (while I still think, that's pretty slow for the amount of Import I requested)
But now the current Jobs take 46min for 1881 products.

Has anyone an idea of where the bottleneck might be?

**Comment by atrolex** [comment]
Hi XCame,
"product" is a very specific entity, with a lot of rules, checks and virtualizations, that is why it takes so long. I don't think we can improve it significantly now.

We know it and that is why we want to migrate everything to PostgreSQL, here you need to wait ca. 3 months.

Currently, for a large amount of data the best practice is to import products and attribute values separately, as "product attribute value" is technically a separate entity in the system. 

Entity "Product attribute value" expects the data in such structure:
ProductID | AttributeName | Attribute Value

If you want to have your attribute values as different columns, you will need to create a separate feed for every attribute.
To improve the usability you can use our module Connector, with this module, you can upload the file only once (for pre-configured connector) and all the feeds linked to this connector will be executed automatically.

Currently we are thinking about the "Transformation" which will automatically convert
ProductID | AttributeName1 | AttributeName2|AttributeName3.... to
ProductID|AttributeName|AttributeValue
so that only one import feed can be configured.

I think this functionality will be added in the Connector module too, when we implement it.

You can also think about not using product attributes, and create all your attributes as fields (will work only if you have not too many attributes.) in the entity "Product".
